clowder-framework/clowder
# Generate a unique jobId for related Extraction messages

## Description
Relates to https://github.com/clowder-framework/pyclowder/pull/7
Fixes #28 

Demonstration: http://recordit.co/BThq8PnxM8

### Problem
As an extractor developer or user, it would be nice to be able to track which jobs generated which status messages. It is difficult to track and debug using ExtractorMessages because for a given file + extractor, if multiple "jobs" or "runs" of the extractor are present, it is hard to determine which messages came from which jobs.

### Approach
* Generate a unique/random UUID when submitting a file to an extractor. We can then forward this jobId to pyClowder, who can send it back in subsequent status updates. Ultimately, this is saved in the database in a new field.
* Rework the Extractions tab in the UI to categorize by jobId
* Sort newest to oldest for top 2 levels (extractors and jobs), but sort oldest to newest for actual event list for each job

### How to Test
1. Clone my pyclowder branch locally and run ./docker.sh to build the images
    * `git clone https://github.com/clowder-framework/pyclowder pyclowder-testing
    * `cd pyclowder-testing/`
    * `./docker.sh`
2. Run the wordcount extractor from the custom pyclowder image you just built
    * `docker run -it --rm --net=host -e REGISTRATION_ENDPOINTS="${REGISTRATION_ENDPOINTS}" -e RABBITMQ_URI="${RABBITMQ_URI}" clowder/extractors-wordcount:latest`
    * NOTE: You should not see Docker pull an image when running this command
3. Start up RabbitMQ + enable RabbitMQ plugin in Clowder config
4. Checkout and run this branch locally
5. Login to Clowder
6. Create a dataset, if you haven't already, and upload a text file to it
    * You can use `echo 'hello world' > hello.json` to create a simple text file
    * You should see the extractor begin to process the file
    * You should see in the extractor logs that check_message outputted a `jobid`
7. Open a mongo shell and query the contents of the `extractions` by job_id:
    * `docker exec -it mongo bash`
    * `mongo`
    * `use clowder`
    * `db.extractions.find({ 'job_id': ObjectId('YOUR_JOBID_HERE') })
    * You should see that only extractions associated with this particular "run" are shown (on this particular extractor + file)
8. View the file in Clowder, check the Extractions tab, and expand an extractor
    * You should see a list of all extractors / events that have run on the file are listed here
    * Upon expanding, you should see a list of clickable unique jobs shown here - note that the last update time is displayed instead of the Job ID
    * You should see a chevron icon point to the right of each job indicating that an expandable row is collapsed
9. Click a job to expand it
    * You should see the chevron icon point downward to indicate the row is expanded
    * You should see all events from that job

## Review Time Estimate


## Types of changes


## Checklist:
